File: DataStuctures/F_StackUsingTwoQueuesCustom.java
/*
Structure:
Queue 1: [ 1 | 2 | 3 ]
Queue 2: [           ]
(To push, add to non-empty queue. To pop, move all but last to other queue)
*/

import java.util.LinkedList;
import java.util.Queue;

/**
 * Custom Stack Implementation using two Queues.
 * 
 * @param <T> the type of elements in this stack
 */
public class F_StackUsingTwoQueuesCustom<T> {
    private Queue<T> q1 = new LinkedList<>();
    private Queue<T> q2 = new LinkedList<>();

    public void push(T data) {
        q2.add(data);
        while (!q1.isEmpty()) {
            q2.add(q1.remove());
        }
        Queue<T> temp = q1;
        q1 = q2;
        q2 = temp;
    }

    public T pop() {
        if (isEmpty()) return null;
        return q1.remove();
    }

    public T peek() {
        if (isEmpty()) return null;
        return q1.peek();
    }

    public boolean isEmpty() {
        return q1.isEmpty();
    }

    public int size() {
        return q1.size();
    }

    public static void main(String[] args) {
        F_StackUsingTwoQueuesCustom<Integer> s = new F_StackUsingTwoQueuesCustom<>();
        s.push(1);
        s.push(2);
        s.push(3);
        System.out.println("Popped: " + s.pop());
        System.out.println("Peek: " + s.peek());
        s.push(4);
        System.out.println("Popped: " + s.pop());
    }
}

File: DataStructures/f_stack_using_two_queues_custom.py
"""
Structure:
Queue 1: [ 1 | 2 | 3 ]
Queue 2: [           ]
(To push, add to non-empty queue. To pop, move all but last to other queue)
"""

from collections import deque

class StackUsingTwoQueuesCustom:
    """Custom Stack Implementation using two Queues."""
    def __init__(self):
        self.q1 = deque()
        self.q2 = deque()

    def push(self, data):
        self.q2.append(data)
        while self.q1:
            self.q2.append(self.q1.popleft())
        self.q1, self.q2 = self.q2, self.q1

    def pop(self):
        if self.is_empty():
            return None
        return self.q1.popleft()

    def peek(self):
        if self.is_empty():
            return None
        return self.q1[0]

    def is_empty(self):
        return len(self.q1) == 0

    def size(self):
        return len(self.q1)


if __name__ == "__main__":
    s = StackUsingTwoQueuesCustom()
    s.push(1)
    s.push(2)
    s.push(3)
    print("Popped:", s.pop())
    print("Peek:", s.peek())
    s.push(4)
    print("Popped:", s.pop())
